src/ProtoBuf.FSharp/CodeGen.fs

```diff
@@ -66,99 +119,381 @@ let private emitFactory (resultType : Type) (zeroValuesPerField: ZeroValues.Fiel
-            let ctr = surrogateType.DefineConstructor(MethodAttributes.Public, CallingConventions.Standard, [| |])
-            let gen = ctr.GetILGenerator()
-            gen.Emit(OpCodes.Ldarg_0)
-            gen.Emit(OpCodes.Call, typeof<obj>.GetConstructor [||])
-            gen.Emit(OpCodes.Ret)
-            ValueSome ctr
+        else surrogateType.DefineDefaultConstructor MethodAttributes.Public |> ValueSome
 
-    let attr = MethodAttributes.Public ||| MethodAttributes.HideBySig ||| MethodAttributes.SpecialName ||| MethodAttributes.Static
+    let struct (constructor, extractMethod) =
+        emitSurrogateContent surrogateType recordType
+            (FSharpType.GetRecordFields(recordType, true))
+            (FSharpValue.PreComputeRecordConstructorInfo(recordType, true))
+            false null
 
-    // Define op_Implicit methods that Protobuf calls to create recordType from surrogate.
-    let conv = surrogateType.DefineMethod("op_Implicit", attr, recordType, [| surrogateType |])
+    // Define op_Explicit methods that Protobuf calls to create recordType from surrogate.
+    let conv = surrogateType.DefineOpExplicit(surrogateType, recordType)
     let gen = conv.GetILGenerator()
-    let ctr = FSharpValue.PreComputeRecordConstructorInfo(recordType, true)
-    for (recordField, surrogateField) in surrogateFields do
-        gen.Emit((if surrogateType.IsValueType then OpCodes.Ldarga_S else OpCodes.Ldarg), 0)
-        gen.Emit(OpCodes.Ldfld, surrogateField)
-        ZeroValues.getZeroValueMethodInfoOpt recordField.PropertyType |> Option.iter (fun getValue ->
-            let skip = gen.DefineLabel()
-            gen.Emit(OpCodes.Dup)
-            gen.Emit(OpCodes.Brtrue, skip)
-            gen.Emit(OpCodes.Pop)
-            emitZeroValueOntoEvaluationStack gen getValue
-            gen.MarkLabel(skip)
-        )
-    gen.Emit(OpCodes.Newobj, ctr)
+    gen.Emit((if surrogateType.IsValueType then OpCodes.Ldarga_S else OpCodes.Ldarg), 0)
+    gen.Emit(OpCodes.Call, extractMethod)
     gen.Emit(OpCodes.Ret)
 
-    // Define op_Implicit methods that Protobuf calls to create surrogate from recordType.
-    let conv = surrogateType.DefineMethod("op_Implicit", attr, surrogateType, [| recordType |])
+    // Define op_Explicit methods that Protobuf calls to create surrogate from recordType.
+    let conv = surrogateType.DefineOpExplicit(recordType, surrogateType)
     let gen = conv.GetILGenerator()
+    let argIsNotNull = gen.DefineLabel()
+    if not recordType.IsValueType then
+        gen.Emit(OpCodes.Ldarg_0)
+        gen.Emit(OpCodes.Brtrue, argIsNotNull)
+        match defaultConstructor with
+        | ValueSome ctr -> gen.Emit(OpCodes.Newobj, ctr)
+        | ValueNone -> emitDefaultValueViaCell gen surrogateType
+        gen.Emit(OpCodes.Ret)
+    gen.MarkLabel(argIsNotNull)
+    gen.Emit((if recordType.IsValueType then OpCodes.Ldarga_S else OpCodes.Ldarg), 0)
+    gen.Emit(OpCodes.Newobj, constructor)
+    gen.Emit(OpCodes.Ret)
```

Comment:
This block above really needs some sort of equivalent code commentary at the very least. Its quite a substantially large PR at this point.